noble-migration/src/bin/check.rs

```diff
@@ -0,0 +1,259 @@
+    fs,
+    path::Path,
+    process::{self, ExitCode},
+};
+use url::{Host, Url};
+
+/// This file contains the state of the pre-migration checks.
+///
+/// There are four possible states:
+/// * does not exist: check script hasn't run yet
+/// * empty JSON object: script determines it isn't on focal
+/// * {"error": true}: script encountered an error
+/// * JSON object with boolean values for each check (see `State` struct)
+const STATE_PATH: &str = "/etc/securedrop-noble-migration.json";
+
+#[derive(Serialize)]
+struct State {
+    ssh: bool,
+    ufw: bool,
+    free_space: bool,
+    apt: bool,
+    systemd: bool,
+}
+
+impl State {
+    fn is_ready(&self) -> bool {
+        self.ssh && self.ufw && self.free_space && self.apt && self.systemd
+    }
+}
+
+/// Parse the OS codename from /etc/os-release
+fn os_codename() -> Result<String> {
+    let contents = fs::read_to_string("/etc/os-release")
+        .context("reading /etc/os-release failed")?;
+    for line in contents.lines() {
+        if line.starts_with("VERSION_CODENAME=") {
+            // unwrap: Safe because we know the line contains "="
+            let (_, codename) = line.split_once("=").unwrap();
+            return Ok(codename.trim().to_string());
+        }
+    }
+
+    bail!("Could not find VERSION_CODENAME in /etc/os-release")
+}
+
+/// Check that the UNIX "ssh" group has no members
+///
+/// See <https://github.com/freedomofpress/securedrop/issues/7316>.
+fn check_ssh_group() -> Result<bool> {
+    // There are no clean bindings to getgrpname in rustix,
+    // so jut shell out to getent to get group members
+    let output = process::Command::new("getent")
+        .arg("group")
+        .arg("ssh")
+        .output()
+        .context("spawning getent failed")?;
+    if output.status.code() == Some(2) {
+        println!("ssh: group does not exist");
+        return Ok(true);
+    } else if !output.status.success() {
+        bail!(
+            "running getent failed: {}",
+            String::from_utf8_lossy(&output.stderr)
+        );
+    }
+
+    let stdout = String::from_utf8(output.stdout)
+        .context("getent stdout is not utf-8")?;
+    // The format looks like `ssh:x:123:member1,member2`
+    let (_, members) = stdout.rsplit_once(':').unwrap();
+    if members.is_empty() {
+        println!("ssh: group is empty");
+        Ok(true)
+    } else {
+        println!("ssh: group is not empty: {members}");
+        Ok(false)
+    }
+}
+
+/// Check that ufw was removed
+///
+/// See <https://github.com/freedomofpress/securedrop/issues/7313>.
+fn check_ufw_removed() -> bool {
+    if Path::new("/usr/sbin/ufw").exists() {
+        println!("ufw: ufw is still installed");
+        false
+    } else {
+        println!("ufw: ufw was removed");
+        true
+    }
+}
+
+const DESIRED_FREE_SPACE: u64 = 10 * 1024 * 1024 * 1024; // 10GB
+
+/// Check that there is enough free space for taking a backup
+/// and downloading all the new packages. 10GB is a bit of an
```

Comment:
I mostly made this number up. IIRC it was like 2-3GB of packages that needed to be downloaded, let's conservatively round it up to 4GB. That leaves 6GB for a backup.

This totally fails in two obvious cases:
1. Instance that has like 500MB of submissions stored and has 8GB free. They have enough space for the upgrade plus backup with enough headroom, but we'll still flag it.
2. Instance with 20GB of submissions stored and 12GB free. They don't have enough space for the backup but we don't flag it.

My current feeling is that we can make this check smarter (and also the backup script) with not too much effort, but let's do it as a follow-up.

Reply:
> Any reason not to do it now

My initial intent was that this PR was just going to be the framework and then we could add more checks iteratively, but we're hitting end-of-sprint and also I haven't really come up with any other checks, so yeah, let me just do it now.